File: example-objc/example-objc/CardFormViewExampleViewController.m
//
//  CardFormViewExampleViewController.m
//  example-objc
//
//  Created by Li-Hsuan Chen on 2019/07/23.
//  Copyright © 2019 PAY, Inc. All rights reserved.
//

#import "CardFormViewExampleViewController.h"
#import "ColorTheme.h"
#import "UIViewController+Alert.h"
@import PAYJP;

@interface CardFormViewExampleViewController ()

@property(weak, nonatomic) IBOutlet PAYCardFormTableStyledView *cardFormView;
@property(weak, nonatomic) IBOutlet UILabel *tokenIdLabel;
@property(weak, nonatomic) IBOutlet UITableViewCell *createTokenCell;
@property(weak, nonatomic) IBOutlet UITableViewCell *validateAndCreateTokenCell;
@property(weak, nonatomic) IBOutlet UITextField *selectColorField;

- (IBAction)cardHolderSwitchChanged:(id)sender;

@property(strong, nonatomic) PAYAPIClient *payjpClient;

@property(strong, nonatomic) NSArray *list;
@property(strong, nonatomic) UIPickerView *pickerView;
@property(assign, nonatomic) PAYTokenOperationStatus tokenOperationStatus;

@end

@implementation CardFormViewExampleViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  self.cardFormView.delegate = self;
  [self fetchBrands];

  self.list = @[ @"Normal", @"Red", @"Blue", @"Dark" ];
  self.pickerView = [[UIPickerView alloc] init];
  self.pickerView.delegate = self;
  self.pickerView.dataSource = self;
  self.pickerView.showsSelectionIndicator = YES;
  self.selectColorField.delegate = self;

  UIToolbar *toolbar = [[UIToolbar alloc] init];
  UIBarButtonItem *spaceItem =
      [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace
                                                    target:self
                                                    action:nil];
  UIBarButtonItem *doneItem =
      [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone
                                                    target:self
                                                    action:@selector(colorSelected:)];
  [toolbar setItems:@[ spaceItem, doneItem ]];
  [toolbar sizeToFit];

  self.selectColorField.inputView = self.pickerView;
  self.selectColorField.inputAccessoryView = toolbar;

  [self toggleCellStyle:self.createTokenCell isEnabled:NO];

  // processing to adjust the cell height of CardFormView when OS version is lower than 10
  [self.tableView layoutIfNeeded];

  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(handleTokenOperationStatusChange:)
                                               name:NSNotification.payjpTokenOperationStatusChanged
                                             object:nil];
}

- (void)colorSelected:(id)sender {
  [self.selectColorField endEditing:YES];
  NSString *selected = self.list[[self.pickerView selectedRowInComponent:0]];
  self.selectColorField.text = selected;
  ColorTheme theme = GetColorTheme(selected);

  switch (theme) {
    case Red: {
      UIColor *red = RGB(255, 69, 0);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:nil
                                                                  inputTextColor:red
                                                                  errorTextColor:nil
                                                                       tintColor:red
                                                       inputFieldBackgroundColor:nil
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = UIColor.clearColor;
      break;
    }
    case Blue: {
      UIColor *blue = RGB(0, 103, 187);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:nil
                                                                  inputTextColor:blue
                                                                  errorTextColor:nil
                                                                       tintColor:blue
                                                       inputFieldBackgroundColor:nil
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = UIColor.clearColor;
      break;
    }
    case Dark: {
      UIColor *white = UIColor.whiteColor;
      UIColor *darkGray = RGB(61, 61, 61);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:nil
                                                                  inputTextColor:white
                                                                  errorTextColor:nil
                                                                       tintColor:white
                                                       inputFieldBackgroundColor:nil
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = darkGray;
      break;
    }
    default: {
      UIColor *black = UIColor.blackColor;
      UIColor *defaultBlue = RGB(0, 122, 255);
      PAYCardFormStyle *style = [[PAYCardFormStyle alloc] initWithLabelTextColor:nil
                                                                  inputTextColor:black
                                                                  errorTextColor:nil
                                                                       tintColor:defaultBlue
                                                       inputFieldBackgroundColor:nil
                                                               submitButtonColor:nil
                                                                  highlightColor:nil];
      [self.cardFormView applyWithStyle:style];
      self.cardFormView.backgroundColor = UIColor.clearColor;
      break;
    }
  }
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
  NSString *sectionName;
  switch (section) {
    case 0:
      sectionName = NSLocalizedString(@"example_card_information_section", nil);
      break;
    case 2:
      sectionName = NSLocalizedString(@"example_token_id_section", nil);
      break;
    default:
      break;
  }
  return sectionName;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  [tableView deselectRowAtIndexPath:indexPath animated:YES];

  // Create Token
  if (indexPath.section == 1 && indexPath.row == 1) {
    if (!self.cardFormView.isValid) {
      return;
    }
    [self createToken];
  }
  // Valdate and Create Token
  if (indexPath.section == 1 && indexPath.row == 2) {
    BOOL isValid = [self.cardFormView validateCardForm];
    if (isValid) {
      [self createToken];
    }
  }
}

- (void)createToken {
  __weak typeof(self) wself = self;

  [self.cardFormView
      createTokenWith:nil
           completion:^(PAYToken *token, NSError *error) {
             if (error.domain == PAYErrorDomain && error.code == PAYErrorServiceError) {
               id<PAYErrorResponseType> errorResponse = error.userInfo[PAYErrorServiceErrorObject];
               NSLog(@"[errorResponse] %@", errorResponse.description);
             }

             if (!token) {
               dispatch_async(dispatch_get_main_queue(), ^{
                 wself.tokenIdLabel.text = nil;
                 [wself showError:error];
               });
               return;
             }

             NSLog(@"token = %@", [wself displayToken:token]);
             dispatch_async(dispatch_get_main_queue(), ^{
               wself.tokenIdLabel.text = token.identifer;
               [wself.tableView reloadData];
               [wself showToken:token];
             });
           }];
}

- (void)fetchBrands {
  __weak typeof(self) wself = self;

  [self.cardFormView
      fetchBrandsWith:@"tenant_id"
           completion:^(NSArray<NSString *> *cardBrands, NSError *error) {
             if (error.domain == PAYErrorDomain && error.code == PAYErrorServiceError) {
               id<PAYErrorResponseType> errorResponse = error.userInfo[PAYErrorServiceErrorObject];
               NSLog(@"[errorResponse] %@", errorResponse.description);
             }

             if (!cardBrands) {
               dispatch_async(dispatch_get_main_queue(), ^{
                 [wself showError:error];
               });
             }
           }];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
  return UITableViewAutomaticDimension;
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
  return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
  return self.list.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView
             titleForRow:(NSInteger)row
            forComponent:(NSInteger)component {
  return self.list[row];
}

- (BOOL)textField:(UITextField *)textField
    shouldChangeCharactersInRange:(NSRange)range
                replacementString:(NSString *)string {
  return NO;
}

- (void)formInputValidatedIn:(UIView *)cardFormView isValid:(BOOL)isValid {
  [self updateButtonEnabled];
}

- (void)formInputDoneTappedIn:(UIView *)cardFormView {
  [self createToken];
}

- (IBAction)cardHolderSwitchChanged:(UISwitch *)sender {
  [self.cardFormView setCardHolderRequired:sender.isOn];
}

- (void)handleTokenOperationStatusChange:(NSNotification *)notification {
  self.tokenOperationStatus =
      [notification.userInfo[PAYNotificationKey.newTokenOperationStatus] integerValue];
  [self updateButtonEnabled];
}

- (void)updateButtonEnabled {
  BOOL isAcceptable = self.tokenOperationStatus == PAYTokenOperationStatusAcceptable;
  [self toggleCellStyle:self.createTokenCell isEnabled:isAcceptable && self.cardFormView.isValid];
  [self toggleCellStyle:self.validateAndCreateTokenCell isEnabled:isAcceptable];
  NSIndexPath *createButtonRow = [NSIndexPath indexPathForRow:1 inSection:1];
  NSIndexPath *validateAndCreateButtonRow = [NSIndexPath indexPathForRow:2 inSection:1];
  [self.tableView reloadRowsAtIndexPaths:@[ createButtonRow, validateAndCreateButtonRow ]
                        withRowAnimation:UITableViewRowAnimationNone];
}

- (void)toggleCellStyle:(UITableViewCell *)cell isEnabled:(BOOL)isEnabled {
  [cell setUserInteractionEnabled:isEnabled];
  cell.selectionStyle =
      isEnabled ? UITableViewCellSelectionStyleDefault : UITableViewCellSelectionStyleNone;
  cell.contentView.alpha = isEnabled ? 1.0 : 0.5;
}

@end

File: example-swift/example-swift/CardFormViewExampleViewController.swift
//
//  CardFormViewExampleViewController.swift
//  example-swift
//
//  Created by Tadashi Wakayanagi on 2019/09/10.
//

import UIKit
import PAYJP

class CardFormVieExampleViewController: UITableViewController, CardFormViewDelegate,
                                        UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate {

    @IBOutlet private weak var formContentView: UIView!
    @IBOutlet private weak var createTokenCell: UITableViewCell!
    @IBOutlet private weak var validateAndCreateTokenCell: UITableViewCell!
    @IBOutlet private weak var tokenIdLabel: UILabel!
    @IBOutlet private weak var selectColorField: UITextField!

    private var cardFormView: CardFormTableStyledView!
    private var tokenOperationStatus: TokenOperationStatus = .acceptable

    private let list: [ColorTheme] = [.Normal, .Red, .Blue, .Dark]
    private var pickerView: UIPickerView!

    override func viewDidLoad() {

        let x: CGFloat = self.formContentView.bounds.origin.x
        let y: CGFloat = self.formContentView.bounds.origin.y

        let width: CGFloat = self.formContentView.bounds.width
        let height: CGFloat = self.formContentView.bounds.height

        let frame: CGRect = CGRect(x: x, y: y, width: width, height: height)
        cardFormView = CardFormTableStyledView(frame: frame)
        cardFormView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        cardFormView.setCardHolderRequired(true)
        cardFormView.delegate = self

        self.formContentView.addSubview(cardFormView)
        self.toggleCellStyle(cell: self.createTokenCell, isEnabled: false)

        self.pickerView = UIPickerView()
        self.pickerView.delegate = self
        self.pickerView.dataSource = self
        self.pickerView.showsSelectionIndicator = true
        self.selectColorField.delegate = self

        let toolbar = UIToolbar()
        let spaceItem = UIBarButtonItem(
            barButtonSystemItem: UIBarButtonItem.SystemItem.flexibleSpace,
            target: nil,
            action: nil)
        let doneItem = UIBarButtonItem(
            barButtonSystemItem: UIBarButtonItem.SystemItem.done,
            target: self,
            action: #selector(colorSelected(_:)))
        toolbar.setItems([spaceItem, doneItem], animated: true)
        toolbar.sizeToFit()

        self.selectColorField.inputView = self.pickerView
        self.selectColorField.inputAccessoryView = toolbar

        self.fetchBrands()

        // processing to adjust the cell height of CardFormView when OS version is lower than 10
        self.tableView.layoutIfNeeded()

        NotificationCenter.default.addObserver(self,
                                               selector: #selector(handleTokenOperationStatusChange(notification:)),
                                               name: .payjpTokenOperationStatusChanged,
                                               object: nil)
    }

    @objc private func colorSelected(_ sender: UIButton) {
        self.selectColorField.endEditing(true)
        let theme = self.list[self.pickerView.selectedRow(inComponent: 0)]
        self.selectColorField.text = theme.rawValue

        switch theme {
        case .Red:
            let red = UIColor(255, 69, 0)
            let style = FormStyle(inputTextColor: red, tintColor: red)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = .clear
        case .Blue:
            let blue = UIColor(0, 103, 187)
            let style = FormStyle(inputTextColor: blue, tintColor: blue)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = .clear
        case .Dark:
            let darkGray = UIColor(61, 61, 61)
            let style = FormStyle(inputTextColor: .white, tintColor: .white)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = darkGray
        default:
            let defaultBlue = UIColor(0, 122, 255)
            let style = FormStyle(inputTextColor: .black, tintColor: defaultBlue)
            self.cardFormView.apply(style: style)
            self.cardFormView.backgroundColor = .clear
        }
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        var sectionName: String?
        switch section {
        case 0:
            sectionName = NSLocalizedString("example_card_information_section", tableName: nil, comment: "")
        case 2:
            sectionName = NSLocalizedString("example_token_id_section", tableName: nil, comment: "")
        default:
            break
        }
        return sectionName
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)

        // Create Token
        if indexPath.section == 1 && indexPath.row == 1 {
            if !self.cardFormView.isValid {
                return
            }
            createToken()
        }
        // Valdate and Create Token
        if indexPath.section == 1 && indexPath.row == 2 {
            let isValid = self.cardFormView.validateCardForm()
            if isValid {
                createToken()
            }
        }
    }

    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableView.automaticDimension
    }

    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }

    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return self.list.count
    }

    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return self.list[row].rawValue
    }

    func textField(
        _ textField: UITextField,
        shouldChangeCharactersIn range: NSRange,
        replacementString string: String
    ) -> Bool {
        return false
    }

    func formInputValidated(in cardFormView: UIView, isValid: Bool) {
        updateButtonEnabled()
    }

    func formInputDoneTapped(in cardFormView: UIView) {
        self.createToken()
    }

    @IBAction func cardHolderSwitchChanged(_ sender: UISwitch) {
        self.cardFormView.setCardHolderRequired(sender.isOn)
        self.tableView.reloadData()
    }

    @objc private func handleTokenOperationStatusChange(notification: Notification) {
        if let value = notification.userInfo?[PAYNotificationKey.newTokenOperationStatus] as? Int,
           let newStatus = TokenOperationStatus.init(rawValue: value) {
            self.tokenOperationStatus = newStatus
            self.updateButtonEnabled()
        }
    }

    private func updateButtonEnabled() {
        let isAcceptable = self.tokenOperationStatus == .acceptable
        self.toggleCellStyle(cell: self.createTokenCell, isEnabled: isAcceptable && self.cardFormView.isValid)
        self.toggleCellStyle(cell: self.validateAndCreateTokenCell, isEnabled: isAcceptable)
        let indexPathCreateToken = IndexPath(row: 1, section: 1)
        let indexPathValidateAndCreateToken = IndexPath(row: 2, section: 1)
        self.tableView.reloadRows(at: [indexPathCreateToken, indexPathValidateAndCreateToken],
                                  with: .none)
    }

    private func toggleCellStyle(cell: UITableViewCell, isEnabled: Bool) {
        cell.isUserInteractionEnabled = isEnabled
        cell.selectionStyle = isEnabled ? .default : .none
        cell.contentView.alpha = isEnabled ? 1.0 : 0.5
    }

    func createToken() {
        self.cardFormView.createToken(tenantId: "tenant_id") { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(let token):
                DispatchQueue.main.async {
                    self.tokenIdLabel.text = token.identifer
                    self.tableView.reloadData()
                    self.showToken(token: token)
                }
            case .failure(let error):
                if let apiError = error as? APIError, let payError = apiError.payError {
                    print("[errorResponse] \(payError.description)")
                }

                DispatchQueue.main.async {
                    self.tokenIdLabel.text = nil
                    self.showError(error: error)
                }
            }
        }
    }

    func fetchBrands() {
        self.cardFormView.fetchBrands(tenantId: "tenant_id") { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(let brands):
                print("card brands => \(brands)")
            case .failure(let error):
                if let payError = error.payError {
                    print("[errorResponse] \(payError.description)")
                }

                DispatchQueue.main.async {
                    self.tokenIdLabel.text = nil
                    self.showError(error: error)
                }
            }
        }
    }
}
